app/services/concerns/foreman_puppet/extensions/bulk_hosts_manager.rb

```diff
@@ -0,0 +1,26 @@
+module ForemanPuppet
+  module Extensions
+    module BulkHostsManager
+      extend ActiveSupport::Concern
+
+      def change_puppet_proxy(proxy_id, ca_proxy)
+        proxy = SmartProxy.find_by(id: proxy_id)
+        begin
+          @hosts.each do |host|
+            if ca_proxy
+              host.puppet_ca_proxy = proxy
+            else
+              host.puppet_proxy = proxy
+            end
+            host.save(:validate => false)
+        rescue StandardError => e
+            message = format(_('Failed to set %{proxy_type} proxy for %{host}.'), host: host, proxy_type: proxy_type)
+            Foreman::Logging.exception(message, e)
+        end
+        rescue RecordNotFoundError => e
+          Foreman::Logging.exception(_('Cound not find Smart Proxy'))
+        end
```

Comment:
```suggestion
        rescue StandardError => e
            message = format(_('Failed to set %{proxy_type} proxy for %{host}.'), host: host, proxy_type: proxy_type)
            Foreman::Logging.exception(message, e)
        end
        rescue ActiveRecord::RecordNotFound => e
          Foreman::Logging.exception(_('Cound not find Smart Proxy'))
        end
```

And even then, ActiveRecord::RecordNotFound is a subclass of StandardError so the record not found branch would never match

Reply:
Yeah, that one is messy - this also somehow depends if we want a separate method for unassign.